src/net_processing.cpp

```diff
@@ -3953,11 +3953,55 @@ void PeerManager::ConsiderEviction(CNode& pto, int64_t time_in_seconds)
 
 void PeerManager::EvictExtraOutboundPeers(int64_t time_in_seconds)
 {
-    // Check whether we have too many outbound peers
-    int extra_peers = m_connman.GetExtraOutboundCount();
-    if (extra_peers > 0) {
-        // If we have more outbound peers than we target, disconnect one.
-        // Pick the outbound peer that least recently announced
+    // If we have any extra block-relay-only peers, disconnect the youngest unless
+    // it's given us a block -- in which case, compare with the second-youngest, and
+    // out of those two, disconnect the peer who least recently gave us a block.
```

Comment:
A peer would never provide us a block if we provided them a block earlier.

So, an attacker can try being veeeery fast at sending blocks to a victim (I assume it's possible) so that other peers are worst, and slowly evict them one by one.
This of course assumes an attacker has eviction capabilities.

Current approach of this PR is more conservative and wouldn't allow this.

Reply:
Essentially this means that we have one 'fixed' block-relay-only peer (even across restarts due to anchor connections) and one evictable block-relay-only peer.

You may be right that this makes it more difficult for an adversary to take over both block-relay-only connections, although it seems like it'd be very difficult to pull off - the adversary would need to take over a large part of the victim's addrman, be lucky with timings of blocks and additional block-relay-only connections, and be sufficiently well connected to the network that they can always win a block race against *all* of the victim's other connections.

In any case, if this is the model (one fixed and one evictable block-relay-only peer), then it'd be much better to explicitly document that rather than it be a consequence of the logic here.

Reply:
I thought I did document this in`net.h`, but I think it would make sense to put a writeup of how all this works on our wiki, which I'd be happy to do after this is merged.

Reply:
I've updated the documentation on the wiki:
https://github.com/bitcoin-core/bitcoin-devwiki/wiki/P2P-Design-Philosophy/_compare/452d8a0c96503ac7158a975acf7d55cd674a3735

Please let me know if you think anything is missing and I'd be happy to clarify further.